### PR TITLE
crimson: block io for recovery ---- recovery part 2

### PR DESCRIPTION
**Pull request by xxhdx1985126**
This PR addresses the issue that client io needs to wait for the recovery of the
corresponding object.

Signed-off-by: Xuehan Xu <xxhdx1985126@163.com>
## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by xxhdx1985126** [comment]
Now the code should be good for review

File: src/crimson/common/exception.h [hunk 1]
```diff
@@ -18,8 +18,21 @@ class system_shutdown_exception final : public std::exception{
   }
 };
 
+class crushmap_changed final : public std::exception{
```
Comment (tchaikov) [inline src/crimson/common/exception.h:21]: i don't think this is accurate. as the acting set is changed, for instance, when an osd is marked down. which is not tracked by the crush map. so probably, s/crushmap/actingset/ ?

Comment (tchaikov) [reply src/crimson/common/exception.h:21]: also, please add a space before "{".

File: src/crimson/common/exception.h [hunk 2]
```diff
@@ -18,8 +18,21 @@ class system_shutdown_exception final : public std::exception{
...
+public:
+  crushmap_changed(bool sp) : still_primary(sp) {}
+  const char* what() const noexcept final {
+    return "acting set changed";
+  }
+  bool is_primary() {
```
Comment (tchaikov) [inline src/crimson/common/exception.h:27]: mark this method `const`?

File: src/crimson/common/exception.h [hunk 3]
```diff
@@ -18,8 +18,21 @@ class system_shutdown_exception final : public std::exception{
...
+    return still_primary;
+  }
+private:
+  bool still_primary = false;
```
Comment (tchaikov) [inline src/crimson/common/exception.h:31]: could mark this member variable `const`. as it never changes after being set in the ctor.

File: src/crimson/osd/osd_operations/client_request.cc [hunk 4]
```diff
@@ -59,38 +59,53 @@ seastar::future<> ClientRequest::start()
...
-	    }
-	    if (is_pg_op()) {
-	      return process_pg_op(pgref);
-	    } else {
-	      return process_op(pgref);
-	    }
-	  });
-	});
+	      handle.enter(pp(pg).await_map)
+	    ).then([this, &pg]() mutable {
```
Comment (tchaikov) [inline src/crimson/osd/osd_operations/client_request.cc:79]: [off topic] why can we assume that pg is still available in this continuation. as we are merely capturing a reference of it.

File: src/crimson/osd/osd_operations/client_request.cc [hunk 5]
```diff
@@ -59,38 +59,53 @@ seastar::future<> ClientRequest::start()
...
+	      }
+	      if (is_pg_op()) {
+		return process_pg_op(pgref);
+	      } else {
+		return process_op(pgref);
+	      }
+	    });
+      }).then([] {
+	return seastar::stop_iteration::yes;
+      }).handle_exception([](std::exception_ptr eptr) {
```
Comment (tchaikov) [inline src/crimson/osd/osd_operations/client_request.cc:99]: why not `handle_exception_type()` directly?

File: src/crimson/osd/osd_operations/client_request.cc (hunk 4 above)
Comment (xxhdx1985126) [reply src/crimson/osd/osd_operations/client_request.cc:79]: I moved pgref into a later lambda, so I think this pg should be available within the lifetime of this lambda. Am I right?

**Comment by xxhdx1985126** [comment]
@tchaikov Done:-)

File: src/crimson/osd/osd.cc [hunk 6]
```diff
@@ -1036,6 +1046,15 @@ seastar::future<> OSD::handle_rep_op_reply(crimson::net::Connection* conn,
   return seastar::now();
 }
 
+seastar::future<> OSD::handle_mark_me_down(crimson::net::Connection* conn,
```
Comment (tchaikov) [inline src/crimson/osd/osd.cc:1095]: why do we need to handle `MOSDMarkMeDown` on the OSD side? i think this message is supposed to be handled by monitor.

File: src/crimson/osd/osd.cc [hunk 7]
```diff
@@ -1180,4 +1199,28 @@ blocking_future<Ref<PG>> OSD::wait_for_pg(
   return pg_map.get_pg(pgid).first;
 }
 
+seastar::future<> OSD::prepare_to_stop()
+{
+  if (osdmap && osdmap->is_up(whoami)) {
+    state.set_prestop();
+    return seastar::with_timeout(
+      seastar::timer<>::clock::now() + std::chrono::seconds(5),
```
Comment (tchaikov) [inline src/crimson/osd/osd.cc:1253]:
```suggestion
    const auto timeout =
      std::chrono::duration_cast<std::chrono::milliseconds>(
        std::chrono::duration<double>(
          local_conf().get_val<double>("osd_mon_shutdown_timeout")));
    return seastar::with_timeout(
      seastar::steady_clock_type::now() + timeout,
```

File: src/crimson/osd/osd.cc [hunk 8]
```diff
@@ -1180,4 +1199,28 @@ blocking_future<Ref<PG>> OSD::wait_for_pg(
...
+      monc->send_message(
+	  make_message<MOSDMarkMeDown>(
+	    monc->get_fsid(),
+	    whoami,
+	    osdmap->get_addrs(whoami),
+	    osdmap->get_epoch(),
+	    true)).then([this] { return stop_acked.get_future(); })
+    ).handle_exception([this](auto&& eptr) {
+      if (state.is_prestop()) {
```
Comment (tchaikov) [inline src/crimson/osd/osd.cc:1262]: i think a better and probably simpler implementation is something like:
```c++
  if (osdmap && osdmap->is_up(whoami)) {
    state.set_prestop();
    return monc->send_message(
      make_message<MOSDMarkMeDown>(
            monc->get_fsid(),
            whoami,
            osdmap->get_addrs(whoami),
            osdmap->get_epoch(),
            true)).then([this] {
    const auto timeout =
      std::chrono::duration_cast<std::chrono::milliseconds>(
        std::chrono::duration<double>(
          local_conf().get_val<double>("osd_mon_shutdown_timeout")));
    return seastar::with_timeout(
      seastar::steady_clock_type::now() + timeout,
      stop_acked.get_future());
    }).handle_exception_type([](seastar::timed_out_error&) {
      // ignore
    });
  } else {
    return seastar::now();
  }
```

in which,

1. move `send_message()` call out of timeout, as we should not take the time for sending a message into the timeout.
1. `assert(state.is_prestop())` as prestop should always set at that moment
1. move `state.set_stopping();` into the `else` branch for better readability
1. check for `seastar::timed_out_error` explicitly.

also, i think there is no need to call `state.set_stopping();` in `got_stop_ack()`. probably a better place for calling `state.set_stopping();` is `OSD::stop()`. like:
```c++
seastar::future<> OSD::stop()
{
  logger().info("stop");
  // see also OSD::shutdown()
  return prepare_to_stop().then([this] {
    state.set_stopping();
    logger().debug("prepared to stop");
```

File: src/crimson/osd/state.h [hunk 9]
```diff
@@ -15,6 +15,7 @@ class OSDState {
     PREBOOT,
     BOOTING,
     ACTIVE,
+    PRESTOP,
```
Comment (tchaikov) [inline src/crimson/osd/state.h:18]:
```suggestion
    PREPARING_TO_STOP,
```

File: src/crimson/osd/state.h [hunk 10]
```diff
@@ -34,6 +35,9 @@ class OSDState {
   bool is_active() const {
     return state == State::ACTIVE;
   }
+  bool is_prestop() const {
```
Comment (tchaikov) [inline src/crimson/osd/state.h:38]:
```suggestion
  bool is_preparing_to_stop() const {
```

File: src/crimson/osd/pg_recovery.cc [hunk 11]
```diff
@@ -259,8 +259,8 @@ size_t PGRecovery::start_backfill_ops(
   size_t max_to_start,
   std::vector<crimson::osd::blocking_future<>> *out)
 {
-  assert(!pg->get_peering_state().get_backfill_targets().empty());
-
+  if (pg->get_peering_state().get_backfill_targets().empty())
```
Comment (tchaikov) [inline src/crimson/osd/pg_recovery.cc:262]: better off putting the rationale in the commit message, IIUC,  `osd_recovery_max_single_start` is not necessarily used up when `start_backfill_ops()` gets called even if it's the last consumer of this setting.

File: src/crimson/osd/osd.cc [hunk 12]
```diff
@@ -734,6 +734,26 @@ seastar::future<bufferlist> OSD::load_map_bl(epoch_t e)
...
+    } else {
+      break;
+    }
+  }
+  return meta_coll->load_maps(first, i).then(
+    [bls=std::move(bls)](auto&& loaded_maps) mutable {
+    bls.merge(loaded_maps);
+    return std::move(bls);
+  });
+}
```
Comment (tchaikov) [inline src/crimson/osd/osd.cc:755]: probably we could use `map_reduce()` here?
```c++
seastar::future<std::map<epoch_t, bufferlist>> OSD::load_map_bls(epoch_t first,
                                                                 epoch_t last)
{
  return seastar::map_reduce(boost::make_counting_iterator<epoch_t>(first),
                             boost::make_counting_iterator<epoch_t>(last + 1),
                             [this](epoch_t e) {
    if (auto found = map_bl_cache.find(e); found) {
      return seastar::make_ready_future<pair<epoch_t, bufferlist>>(
        std::make_pair(e, std::move(*found)));
    } else {
      return load_map_bl(e).then([e](auto&& bl) {
        return seastar::make_ready_future<pair<epoch_t, bufferlist>>(
          std::make_pair(e, std::move(bl)));
      });
    }
  },
  std::map<epoch_t, bufferlist>{},
  [](auto&& bls, auto&& epoch_bl) {
    bls.emplace(std::move(epoch_bl));
    return std::move(bls);
  });
}
```

File: src/crimson/osd/osd_operations/client_request.cc (hunk 4 above)
Comment (tchaikov) [reply src/crimson/osd/osd_operations/client_request.cc:79]: yeah, i just noticed 
```c++
pgref=std::move(pgref)
```
thanks for explaining this to me!

File: src/crimson/common/exception.h [hunk 13]
```diff
@@ -18,8 +18,21 @@ class system_shutdown_exception final : public std::exception{
   }
 };
 
+class actingset_changed final : public std::exception{
```
Comment (tchaikov) [inline src/crimson/common/exception.h:21]: could you add a space after `std::exception` ?

File: src/crimson/common/exception.h [hunk 14]
```diff
@@ -18,8 +18,21 @@ class system_shutdown_exception final : public std::exception{
...
+  bool is_primary() const {
+    return still_primary;
+  }
+private:
+  const bool still_primary = false;
+};
+
 template<typename Func, typename... Args>
-inline seastar::future<> handle_system_shutdown(Func&& func, Args&&... args)
+inline auto handle_system_shutdown(Func&& func, Args&&... args)
```
Comment (tchaikov) [inline src/crimson/common/exception.h:35]: nit, i still think it'd be nice if the return type could be explicitly noted. it helps the reader to understand the this function better.

File: src/crimson/osd/osd_operations/client_request.cc [hunk 15]
```diff
@@ -59,38 +59,49 @@ seastar::future<> ClientRequest::start()
...
+    return seastar::repeat([this, opref]() mutable {
+      return with_blocking_future(handle.enter(cp().await_map))
+      .then([this]() {
+	return with_blocking_future(osd.osdmap_gate.wait_for_map(m->get_min_epoch()));
+      }).then([this](epoch_t epoch) {
+	return with_blocking_future(handle.enter(cp().get_pg));
+      }).then([this] {
+	return with_blocking_future(osd.wait_for_pg(m->get_spg()));
+      }).then([this, opref](Ref<PG> pgref) {
+	    PG &pg = *pgref;
```
Comment (tchaikov) [inline src/crimson/osd/osd_operations/client_request.cc:71]: wrong indent.

File: src/crimson/osd/pg_backend.h [hunk 16]
```diff
@@ -137,11 +137,18 @@ class PGBackend
...
+    bool is_primary;
+  };
+  virtual void on_actingset_changed(pg_change_info pi) = 0;
+  virtual void on_activate_complete();
 protected:
   const shard_id_t shard;
   CollectionRef coll;
   crimson::os::FuturizedStore* store;
   bool stopping = false;
+  bool peering = false;
```
Comment (tchaikov) [inline src/crimson/osd/pg_backend.h:150]: how about:
```c++
struct peering_info_t {
  bool is_primary;
};
std::optional<peering_info_t> peering;
```
?

the reason why i don't like `pg_change_info pci` in this change is

1. `pci` only makes sense when `peering` is true
1. the variable name of `pci` is difficult to understand at the first glance.

File: src/crimson/osd/replicated_backend.cc [hunk 17]
```diff
@@ -91,6 +100,15 @@ ReplicatedBackend::_submit_transaction(std::set<pg_shard_t>&& pg_shards,
     });
 }
 
+void ReplicatedBackend::on_actingset_changed(pg_change_info pi)
+{
+  peering = true;
+  pci = pi;
+  for (auto& [tid, pending_txn] : pending_trans) {
+    pending_txn.all_committed.set_exception(crimson::common::actingset_changed(pci.is_primary));
```
Comment (tchaikov) [inline src/crimson/osd/replicated_backend.cc:108]: might want to wrap this line? like
```c++
  crimson::common::actingset_changed actingset_changed{pci.is_primary};
  for (auto& [tid, pending_txn] : pending_trans) {
    pending_txn.all_committed.set_exception(actingset_changed);
  }
```

File: src/crimson/osd/osd.cc (hunk 12 above)
Comment (xxhdx1985126) [reply src/crimson/osd/osd.cc:755]: I think this may involve some operations done twice, like map_bl_cache.find, right?

Comment (tchaikov) [reply src/crimson/osd/osd.cc:755]: @xxhdx1985126 could you be more specific? why `map_bl_cache.find()` could be called twice with the same epoch?

Comment (xxhdx1985126) [reply src/crimson/osd/osd.cc:755]: @tchaikov there's another map_bl_cache.find() in OSD::load_map_bl()

Comment (xxhdx1985126) [reply src/crimson/osd/osd.cc:756]: but i think we can skip the map_bl_cache.find() in load_map_bls

File: src/crimson/osd/pg_recovery.cc (hunk 11 above)
Comment (tchaikov) [reply src/crimson/osd/pg_recovery.cc:262]: @xuxuehan could you reconsider this comment?

File: src/crimson/common/exception.h [hunk 18]
```diff
@@ -18,6 +18,19 @@ class system_shutdown_exception final : public std::exception{
...
+public:
+  actingset_changed(bool sp) : still_primary(sp) {}
+  const char* what() const noexcept final {
+    return "acting set changed";
+  }
+  bool is_primary() const {
+    return still_primary;
+  }
+private:
+  const bool still_primary = false;
```
Comment (tchaikov) [inline src/crimson/common/exception.h:31]: i'd suggest remove `= false`. as `still_primary` is always set in the constructor.

File: src/crimson/mon/MonClient.cc [hunk 19]
```diff
@@ -557,7 +557,11 @@ void Client::ms_handle_reset(crimson::net::ConnectionRef conn, bool is_replace)
     } else if (active_con && active_con->is_my_peer(conn->get_peer_addr())) {
       logger().warn("active conn reset {}", conn->get_peer_addr());
       active_con.reset();
-      return reopen_session(-1);
+      return reopen_session(-1)
+      .then([this] {
```
Comment (tchaikov) [inline src/crimson/mon/MonClient.cc:561]: could you move `.then([this] {` up into the previous line?

File: src/crimson/mon/MonClient.cc [hunk 20]
```diff
@@ -801,7 +805,11 @@ seastar::future<> Client::handle_monmap(crimson::net::Connection* conn,
     }
   } else {
     logger().warn("mon.{} went away", cur_mon);
-    return reopen_session(-1);
+    return reopen_session(-1)
+    .then([this] {
```
Comment (tchaikov) [inline src/crimson/mon/MonClient.cc:809]: ditto.

File: src/crimson/mon/MonClient.cc [hunk 21]
```diff
@@ -917,7 +925,11 @@ std::vector<unsigned> Client::get_random_mons(unsigned n) const
 
 seastar::future<> Client::authenticate()
 {
-  return reopen_session(-1);
+  return reopen_session(-1)
+  .then([this] {
```
Comment (tchaikov) [inline src/crimson/mon/MonClient.cc:929]: ditto.

File: src/crimson/mon/MonClient.cc [hunk 22]
```diff
@@ -1024,14 +1036,32 @@ Client::run_command(const std::vector<std::string>& cmd,
...
+  return send_message(m).then([&req] {
     return req.get_future();
   });
 }
 
 seastar::future<> Client::send_message(MessageRef m)
 {
-  return active_con->get_conn()->send(m);
+  if (active_con) {
+    if (pending_messages.size()) {
```
Comment (tchaikov) [inline src/crimson/mon/MonClient.cc:1047]:
```suggestion
    if (!pending_messages.empty()) {
```

File: src/crimson/mon/MonClient.cc [hunk 23]
```diff
@@ -1024,14 +1036,32 @@ Client::run_command(const std::vector<std::string>& cmd,
...
+    }
+    return active_con->get_conn()->send(m);
+  }
+  pending_messages.emplace_back(m);
+  return pending_messages.back().pr.get_future();;
+}
+
+void Client::send_pendings()
+{
+  if (active_con) {
```
Comment (tchaikov) [inline src/crimson/mon/MonClient.cc:1058]: i think we can assume that we always have an `active_con` here.

File: src/crimson/mon/MonClient.cc [hunk 24]
```diff
@@ -1024,14 +1036,32 @@ Client::run_command(const std::vector<std::string>& cmd,
...
 {
-  return active_con->get_conn()->send(m);
+  if (active_con) {
+    if (pending_messages.size()) {
+      send_pendings();
+    }
+    return active_con->get_conn()->send(m);
+  }
+  pending_messages.emplace_back(m);
+  return pending_messages.back().pr.get_future();;
```
Comment (tchaikov) [inline src/crimson/mon/MonClient.cc:1053]: i'd suggest move this block into an `else` clause for better readability.

Comment (tchaikov) [reply src/crimson/mon/MonClient.cc:1053]:
```suggestion
 auto delayed = pending_messages.emplace_back(m);
 return delayed->pr.get_future();;
```

File: src/crimson/mon/MonClient.h [hunk 25]
```diff
@@ -163,6 +164,14 @@ class Client : public crimson::net::Dispatcher,
   std::vector<unsigned> get_random_mons(unsigned n) const;
   seastar::future<> _add_conn(unsigned rank, uint64_t global_id);
   crimson::common::Gated gate;
+
+  // messages that are waiting for the active_con to be available
+  struct delayed_msg_send {
```
Comment (tchaikov) [inline src/crimson/mon/MonClient.h:169]:
```suggestion
  struct pending_msg_t {
```

File: src/crimson/mon/MonClient.cc (hunk 23 above)
Comment (xxhdx1985126) [reply src/crimson/mon/MonClient.cc:1058]: Um...I think there maybe a tiny chance that, between the running of this send_pendings() method and session reopening, the connection gets reset again, is this right?

File: src/crimson/mon/MonClient.cc (hunk 24 above)
Comment (xxhdx1985126) [reply src/crimson/mon/MonClient.cc:1053]: This seems not feasible, seastar::promise can't be copied.

Comment (tchaikov) [reply src/crimson/mon/MonClient.cc:1053]:
```suggestion
 auto& delayed = pending_messages.emplace_back(m);
 return delayed->pr.get_future();;
```

@xxhdx1985126 how about chaning `auto` to `auto&` as above?

File: src/crimson/osd/osd_operations/background_recovery.h [hunk 26]
```diff
@@ -23,24 +23,42 @@ class BackgroundRecovery final : public OperationT<BackgroundRecovery> {
...
+  UrgentRecovery(
+    const hobject_t& soid,
+    const eversion_t& need,
+    Ref<PG> pg,
+    ShardServices& ss,
+    epoch_t epoch_started,
+    crimson::osd::scheduler::scheduler_class_t scheduler_class)
+  : BackgroundRecovery{pg, ss, epoch_started, scheduler_class},
+    soid{soid}, need(need) {}
+  virtual void print(std::ostream&) const override;
```
Comment (tchaikov) [inline src/crimson/osd/osd_operations/background_recovery.h:56]: nit, no need to add `virtual` if the `override` specifier is added.

Comment (tchaikov) [reply src/crimson/osd/osd_operations/background_recovery.h:56]: but i think we'd better s/override/final/.

**Comment by tchaikov** [comment]
@xuxuehan and https://github.com/ceph/ceph/pull/35469#discussion_r439941291

File: src/crimson/osd/pg_recovery.cc (hunk 11 above)
Comment (xxhdx1985126) [reply src/crimson/osd/pg_recovery.cc:262]: > better off putting the rationale in the commit message, IIUC, `osd_recovery_max_single_start` is not necessarily used up when `start_backfill_ops()` gets called even if it's the last consumer of this setting.

Yes, that's right, I'll put it in the commit message:-)

**Pull request merged**